File: leetcode-cn/0725.0_Split_Linked_List_in_Parts.py
'''
T:O(2N), S:O(1)

执行用时：40 ms, 在所有 Python3 提交中击败了49.47% 的用户
内存消耗：15.4 MB, 在所有 Python3 提交中击败了41.64% 的用户
通过测试用例：43 / 43
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: ListNode, k: int) -> List[ListNode]:
        L = self.getLen(head)
        q, r = divmod(L, k)
        ans = []
        nth = 0
        node = head
        while node:
            ans.append(node)
            x = 0
            need = q - 1
            if nth < r:
                need += 1 
            while node and x < need:
                node = node.next
                x += 1
            # split
            next_node = node.next
            node.next = None
            node = next_node
            nth += 1

        ans += [None] * (k - nth)
        return ans

    def getLen(self, node):
        n = 0
        while node:
            n += 1
            node = node.next
        return n


'''
use pre ptr
T:O(2N), S:O(1)

执行用时：36 ms, 在所有 Python3 提交中击败了70.74% 的用户
内存消耗：15.4 MB, 在所有 Python3 提交中击败了64.89% 的用户
通过测试用例：43 / 43
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: ListNode, k: int) -> List[ListNode]:
        L = self.getLen(head)
        q, r = divmod(L, k)
        ans = []
        nth = 0
        pre = node = head
        while node:
            ans.append(node)
            x = 0
            need = q
            if nth < r:
                need += 1 
            while node and x < need:
                pre = node
                node = node.next
                x += 1
            # split
            pre.next = None
            nth += 1

        ans += [None] * (k - nth)
        return ans

    def getLen(self, node):
        n = 0
        while node:
            n += 1
            node = node.next
        return n

File: cpp/0725.0_Split_Linked_List_in_Parts.cpp
/*
T: O(2N)
S:O(1)

执行用时：0 ms, 在所有 C++ 提交中击败了100.00% 的用户
内存消耗：8.7 MB, 在所有 C++ 提交中击败了31.94% 的用户
通过测试用例：43 / 43
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int getLen(ListNode* node) {
        int l = 0;
        while (node) {
            node = node->next;
            l++;
        }
        return l;
    }
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        int l = getLen(head);
        int rmn = l % k;
        l /= k;
        ListNode* p = head;
        ListNode* pre = head;
        vector<ListNode*> ans;
        while (p) {
            ans.push_back(p);
            for (int i = 0; i < l; ++i) {
                pre = p;
                p = p->next;
            }
            if (rmn) {
                rmn--;
                pre = p;
                p = p->next;
            }
            // cut
            pre->next = NULL;
        }

        int empty = k - ans.size();
        for (int i = 0; i < empty; i++) {
            ans.push_back(NULL);
        }
        return ans;
    }
};



